### PR TITLE
IServer: Add SnapNewItem() template to simplify the API, do some small refactoring

### PR DESCRIPTION
**Pull request by Kaffeine**
> What is the motivation for the changes of this pull request?

The idea is to simplify and refactor the game code step by step. This is the next step after a similar https://github.com/ddnet/ddnet/pull/5600

The key idea is that all those
```cpp
protocol7::CNetObj_SpectatorInfo *pSpectatorInfo = static_cast<protocol7::CNetObj_SpectatorInfo *>(Server()->SnapNewItem(NETOBJTYPE_SPECTATORINFO, m_ClientID, sizeof(protocol7::CNetObj_SpectatorInfo)));
```
are too verbose and error-prone.

We have the relation between `CNetObj_` classes and their corresponding `NETOBJTYPE_` so we can use it to remove most of the arguments and the cast.
With some small changes in the generated code (e.g. MsgID leveraged from NetMessage to NetObject) we can come to this:
```cpp
auto *pSpectatorInfo = Server()->SnapNewItem<protocol7::CNetObj_SpectatorInfo>(m_ClientID);
```
Yet, we do not use auto and personally I'm skeptic about `auto` in any non-obvious cases so I still keep the class name on the left side. The best show-case of the new API is:
```diff
@@ -330,7 +330,7 @@ void CGameContext::CreatePlayerSpawn(vec2 Pos, CClientMask Mask)
 void CGameContext::CreateDeath(vec2 Pos, int ClientID, CClientMask Mask)
 {
        // create the event
-       CNetEvent_Death *pEvent = (CNetEvent_Death *)m_Events.Create(NETEVENTTYPE_DEATH, sizeof(CNetEvent_Death), Mask);
+       CNetEvent_Death *pEvent = m_Events.Create<CNetEvent_Death>(Mask);
        if(pEvent)
        {
                pEvent->m_X = (int)Pos.x;
@@ -345,7 +345,7 @@ void CGameContext::CreateSound(vec2 Pos, int Sound, CClientMask Mask)
                return;
 
        // create a sound
-       CNetEvent_SoundWorld *pEvent = (CNetEvent_SoundWorld *)m_Events.Create(NETEVENTTYPE_SOUNDWORLD, sizeof(CNetEvent_SoundWorld), Mask);
+       CNetEvent_SoundWorld *pEvent = m_Events.Create<CNetEvent_SoundWorld>(Mask);
        if(pEvent)
        {
                pEvent->m_X = (int)Pos.x;
@@ -376,7 +376,7 @@ bool CGameContext::SnapLaserObject(const CSnapContext &Context, int SnapID, cons
 {
        if(Context.ClientVersion >= VERSION_DDNET_MULTI_LASER)
        {
-               CNetObj_DDNetLaser *pObj = static_cast<CNetObj_DDNetLaser *>(Server()->SnapNewItem(NETOBJTYPE_DDNETLASER, SnapID, sizeof(CNetObj_DDNetLaser)));
+               CNetObj_DDNetLaser *pObj = Server()->SnapNewItem<CNetObj_DDNetLaser>(SnapID);
                if(!pObj)
                        return false;
 
@@ -390,7 +390,7 @@ bool CGameContext::SnapLaserObject(const CSnapContext &Context, int SnapID, cons
        }
        else
        {
-               CNetObj_Laser *pObj = static_cast<CNetObj_Laser *>(Server()->SnapNewItem(NETOBJTYPE_LASER, SnapID, sizeof(CNetObj_Laser)));
+               CNetObj_Laser *pObj = Server()->SnapNewItem<CNetObj_Laser>(SnapID);
                if(!pObj)
                        return false;
```

The only difficulty is `protocol7` handling but I think I did it right (I had to change the logic for a bit and I still check it while waiting for your comments).

I can extract any of the commits to another PR if wanted.

As for https://github.com/ddnet/ddnet/pull/6330/commits/305670eecb628f40bfbf4fca40d89e460127a3d8 I do not like to use CGameContext for the network stuff but we already have kinda similar helper functions and e.g. the NetObjHandlers there, and it is still a change to the better. **The long-term goal is to move that stuff to a new class** (e.g. some kind of protocol abstraction class or layer).

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

File: src/game/server/gamecontext.h [hunk 1]
```diff
@@ -59,6 +59,17 @@ class IStorage;
...
 
+struct CSnapContext
+{
+	CSnapContext(int Version, bool Sixup = false) :
+		ClientVersion(Version), IsSixup(Sixup)
+	{
+	}
+
+	int ClientVersion;
+	bool IsSixup;
```
Comment (Robyt3) [inline src/game/server/gamecontext.h:70]: Prefix `m_`.

Maybe make the fields private and add getters only, if the values are not supposed to change after the context is created.

Comment (Kaffeine) [reply src/game/server/gamecontext.h:70]: Agree :+1:

File: datasrc/datatypes.py [hunk 2]
```diff
@@ -220,6 +220,7 @@ def emit_declaration(self):
 			lines += [f"struct {self.struct_name} : public {self.base_struct_name}", "{"]
 		else:
 			lines += [f"struct {self.struct_name}", "{"]
+		lines += [f"\tstatic constexpr int s_MsgID = {self.enum_name};"]
```
Comment (Chairn) [inline datasrc/datatypes.py:223]: ms_MsgId as it is a static member

File: datasrc/seven/datatypes.py [hunk 3]
```diff
@@ -222,6 +222,8 @@ def emit_declaration(self):
 			lines = [f"struct {self.struct_name} : public {self.base_struct_name}", "{"]
 		else:
 			lines = [f"struct {self.struct_name}", "{"]
+		lines += ["\tusing is_sixup = char;"]
+		lines += [f"\tstatic constexpr int s_MsgID = {self.enum_name};"]
```
Comment (Chairn) [inline datasrc/seven/datatypes.py:226]: idem

File: src/game/server/gamecontext.h (hunk 1 above)
Comment (Kaffeine) [reply src/game/server/gamecontext.h]: Fixed.

File: datasrc/datatypes.py (hunk 2 above)
Comment (Kaffeine) [reply datasrc/datatypes.py]: Thanks!

File: datasrc/seven/datatypes.py (hunk 3 above)
Comment (Kaffeine) [reply datasrc/seven/datatypes.py]: :heavy_plus_sign:

**Comment by Kaffeine** [comment]
(Fixed only the code style)

**Comment by Robyt3** [comment]
LGTM!

**Comment by def-** [comment]
bors r+
Thanks! Looks cleaner

**Pull request merged**